src/coprocessor/dag/batch/executors/util/hash_aggr_helper.rs

```diff
@@ -42,12 +48,14 @@ impl HashAggregationHelper {
                     }
                 }
                 RpnStackNode::Vector { value, .. } => {
+                    let physical_vec = value.as_ref();
+                    let logical_rows = value.logical_rows();
                     match_template_evaluable! {
-                        TT, match &*value {
-                            VectorValue::TT(vector_value) => {
-                                for (row_index, offset) in states_offset_each_row.iter().enumerate() {
+                        TT, match physical_vec {
+                            VectorValue::TT(vec) => {
+                                for (logical_row_index, offset) in states_offset_each_logical_row.iter().enumerate() {
```

Comment:
I guess we can save some bounds checking by zipping `logical_rows` and `states_offset_each_logical_row` .